### PR TITLE
Volume-coupled flow tutorial

### PR DESCRIPTION
**Pull request by tirgendetwas**
Tutorial to validate volume coupling on a mesh region *and* the reading of internal fields in OpenFOAM. Flow is introduced into the system through a square region of cells. The dummy participant writes Velocity at a constant value, the fluid participant reads that value on a `cellSet` and enforces it through `fvOptions`.

This tutorial could be a basis for a more complex scenario where the dummy solver is replaced or expanded.

TODO:

- [x] README.md

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,10 @@
+{
```
Comment (MakisH) [inline .vscode/settings.json:1]: This file should not be here

File: channel-transport/fluid-openfoam/run.sh [hunk 2]
```diff
@@ -0,0 +1,8 @@
+#!/bin/sh
+set -e -u
+
+blockMesh
+touch fluid-openfoam-pimplefoam.foam
```
Comment (MakisH) [inline channel-transport/fluid-openfoam/run.sh:5]:
```suggestion
touch fluid-openfoam.foam
```

(probably copy-paste from the partitioned pipe tutorial, where we have more than one OpenFOAM cases)

File: channel-transport/fluid-openfoam/system/fvSolution [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+solvers
+{
+
+    p
+    {
+            
+    solver           PCG;
+    preconditioner   DIC;
+    tolerance        1e-8;
+    relTol 	         1.0e-3;
```
Comment (MakisH) [inline channel-transport/fluid-openfoam/system/fvSolution:19]:
```suggestion
    tolerance        1e-8;
    relTol 	     1e-3;
```

File: channel-transport/precice-config.xml [hunk 4]
```diff
@@ -19,17 +19,19 @@
     </mesh>
 
     <participant name="Fluid">
-      <provide-mesh name="Fluid-Mesh" />
-      <receive-mesh name="Transport-Mesh" from="Transport" />
+      <use-mesh name="Fluid-Mesh" provide="true"/>
```
Comment (MakisH) [inline channel-transport/precice-config.xml:22]: This PR reverts to the preCICE v2 config format. We need to keep it in the v3 format.

File: volume-coupled-flow/dummy-nutils/dummy.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+from nutils import function, mesh, cli, solver, export
+import treelog as log
+import numpy as np
+import precice
+from mpi4py import MPI
+
+
+def main():
+
+    print("Running utils")
```
Comment (MakisH) [inline volume-coupled-flow/dummy-nutils/dummy.py:13]:
```suggestion
    print("Running Nutils")
```

File: volume-coupled-flow/dummy-nutils/dummy.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+    # cloud of Gauss points
+    uniform_grid = domain.sample("uniform", degree=1)
+
+    # Nutils namespace
+    ns = function.Namespace(fallback_length=2)
+    ns.x = geom
+
+    # preCICE setup
+    interface = precice.Interface("Dummy-Velocity", "../precice-config.xml", 0, 1)
```
Comment (MakisH) [inline volume-coupled-flow/dummy-nutils/dummy.py:33]: Is "Dummy" the right name? We usually name the participants based on the kind of domain they solve (fluid, solid, etc).

File: volume-coupled-flow/fluid-openfoam/0/U_vol [hunk 7]
```diff
@@ -0,0 +1,40 @@
+FoamFile
+{
+    version     2.0;
+    format      ascii;
+    class       volVectorField;
+    object      U_vol;
```
Comment (MakisH) [inline volume-coupled-flow/fluid-openfoam/0/U_vol:6]: Is this file required, or is it created by the adapter if it does not already exist?
In case of the latter, better remove it.

File: volume-coupled-flow/fluid-openfoam/system/controlDict [hunk 8]
```diff
@@ -0,0 +1,46 @@
+FoamFile
+{
+    version     2.0;
+    format      ascii;
+    class       dictionary;
+    object      controlDict;
+}
+
+application     buoyantPimpleFoam;
```
Comment (MakisH) [inline volume-coupled-flow/fluid-openfoam/system/controlDict:9]: Is this the right solver?
By the way, the case also needs a README.md

File: volume-coupled-flow/fluid-openfoam/system/topoSetDict [hunk 9]
```diff
@@ -0,0 +1,29 @@
+/*--------------------------------*- C++ -*----------------------------------*\
+| =========                 |                                                 |
+| \\      /  F ield         | OpenFOAM: The Open Source CFD Toolbox           |
+|  \\    /   O peration     | Version:  v2212                                 |
+|   \\  /    A nd           | Website:  www.openfoam.com                      |
+|    \\/     M anipulation  |                                                 |
+\*---------------------------------------------------------------------------*/
```
Comment (MakisH) [inline volume-coupled-flow/fluid-openfoam/system/topoSetDict:7]: Remove these lines (and optionally the ones below).

File: volume-coupled-flow/precice-config.xml [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+    <mesh name="Fluid-Mesh">
+      <use-data name="Velocity" />
+    </mesh>
+
+    <mesh name="Dummy-Mesh">
+      <use-data name="Velocity" />
+    </mesh>
+
+    <participant name="Fluid">
+      <use-mesh name="Fluid-Mesh" provide="yes" />
```
Comment (MakisH) [inline volume-coupled-flow/precice-config.xml:22]: Also needs to be ported to preCICE v3.

**Review by MakisH — COMMENTED** [review]
Some first comments for the next iteration. Keep it rolling!

File: volume-coupled-flow/fluid-openfoam/0/U_vol (hunk 7 above)
Comment (tirgendetwas) [reply volume-coupled-flow/fluid-openfoam/0/U_vol:6]: It is required, I'm guessing to give the variable an initial state

File: volume-coupled-flow/fluid-openfoam/system/controlDict (hunk 8 above)
Comment (tirgendetwas) [reply volume-coupled-flow/fluid-openfoam/system/controlDict:9]: Why not? This is the solver we have been experimenting with so far

File: volume-coupled-flow/dummy-nutils/dummy.py (hunk 6 above)
Comment (tirgendetwas) [reply volume-coupled-flow/dummy-nutils/dummy.py]: I thought that participant is so basic that it couldn't really be described in a better way

File: volume-coupled-flow/fluid-openfoam/system/controlDict (hunk 8 above)
Comment (MakisH) [reply volume-coupled-flow/fluid-openfoam/system/controlDict:9]: Since we are not simulating any heat transfer (zero temperature gradients in the domain), pimpleFoam would have been enough. Not really an issue, it just looks strange. I think I will change the case to use pimpleFoam, instead.

File: volume-coupled-flow/dummy-nutils/dummy.py (hunk 6 above)
Comment (MakisH) [reply volume-coupled-flow/dummy-nutils/dummy.py]: Good point... In other cases, the participant name is based on the domain. I will rename this to "Source", I think this will be clearer.

File: volume-coupled-flow/fluid-openfoam/system/controlDict (hunk 8 above)
Comment (tirgendetwas) [reply volume-coupled-flow/fluid-openfoam/system/controlDict:9]: Ok. But I think the case looks different in pimpleFoam, so the screenshot in the README has to change as well.

Comment (MakisH) [reply volume-coupled-flow/fluid-openfoam/system/controlDict:9]: Then let's keep the buoyantPimpleFoam, as it also can be a testbed/starting point for writing temperature.

**Comment by MakisH** [comment]
Status: This case currently stumbles upon this preCICE assertion on the Nutils side:

```
---[precice]  iteration: 1, time-window: 201, time: 1 of 4, time-window-size: 0.005, max-time-step-siz
e: 0.005, ongoing: yes, time-window-complete: yes, 
ASSERTION FAILED
Location:   void precice::cplscheme::BaseCouplingScheme::sendData(const PtrM2N&, const DataMap&, bool)
File:       /home/gc/repos/precice/precice/src/cplscheme/BaseCouplingScheme.cpp:119
Expression: math::equals(timesAscending(nTimeSteps - 1), time::Storage::WINDOW_END)
Rank:       0
Arguments:  
  0: timesAscending(nTimeSteps - 1) == 0.9999999999999787
  1: nTimeSteps == 2
```

which is discussed in https://github.com/precice/precice/issues/1751.

Since:

- The case was previously running correctly with preCICE v2.5.0
- The results we obtain till it hangs (at t=0.35) look reasonable
- It could be a test case to evaluate if the original issue is resolved (@BenjaminRodenberg)

I suggest merging this already.

**Review by MakisH — APPROVED** [review]
